### ISSUE TITLE
[Feature Request] Add kl divergence method for torch.distributions

### ISSUE THREAD

**Issue by p-morais** [pytorch/pytorch#4330]
This could be implemented as ```dist1.kldiv(dist2)``` (provided ```type(dist1) == type(dist2)```) or something similar. KL divergence is a very useful diagnostic for policy gradient algorithms in general and some have a KL divergence penalty as part of their objective function ([e.g.](https://arxiv.org/pdf/1707.02286.pdf)), so I think this would be useful to have.

~~Also, the documentation mentions an [entropy()](http://pytorch.org/docs/master/distributions.html#torch.distributions.Distribution.entropy) method but doing dir(torch.distributions.Distribution) on my installation shows no such method. Is that correct?~~

**Comment by apaszke** [comment]
That's expected. The docs you're looking at are for the `master` branch, and you're likely using the last binary version. `entropy()` has been added in the last week only

**Comment by p-morais** [comment]
Oh, duh, my bad. 

Any thoughts on kl div as part of distributions though? If it's something that sounds reasonable, and someone can help with designing the proper tests, I could submit a PR for it.

**Comment by apaszke** [comment]
I think there's been some discussion about it, but I'm not 100% what was the conclusion (also, I don't think we converged on the API). @fritzo, do you remember?

**Comment by fritzo** [comment]
This was briefly discussed on the [Design Doc](https://docs.google.com/document/d/1wnABg0cdyaVMr-Xqz_brHBwnPJuzTeSk3Oqko2HHSfE/edit#heading=h.7w4tro6t2bxr), but no conclusion was reached. We should discuss more. Here are my thoughts:

It would be great to maximize compatibility with [Tensorflow Distributions](https://www.tensorflow.org/versions/r0.12/api_docs/python/contrib.distributions/kullback-leibler_divergence) and provide a function `torch.distributions.kl(p,q)`. The `kl(p,q)` method seems cleaner to me than a `p.kl(q)` method. I also like the idea of a registration mechanism for multiple dispatch, maybe a decorator like
```py
@torch.distributions.register_kl(Normal, Normal)
def _normal_normal_kl(p, q):
    assert isinstance(p, Normal) and isinstance(q, Normal)
    ...implementation...
```

**Comment by fritzo** [comment]
Also see the issue https://github.com/probtorch/pytorch/issues/37 on https://github.com/probtorch/pytorch/projects/1 where we're tracking distributions work (sorry for duplication, but the separate issue list really helps us prevent issues from getting lost).

**Comment by apaszke** [comment]
I also think the functional form would make a better API.

**Comment by vishwakftw** [comment]
@apaszke I think this can be closed now, given that a framework has been established, and many KL-divergence methods exist.
